lib/src/native/realm_core.dart

```diff
@@ -500,52 +502,45 @@ class _RealmCore {
 
     if (syncError is ClientResetError) {
       syncConfig.clientResetHandler.onManualReset?.call(syncError);
-    } else {
-      syncConfig.syncErrorHandler(syncError);
+      return;
+    }
+
+    syncConfig.syncErrorHandler(syncError);
+  }
+
+  static void _callResetCallback(FutureOr<void> Function() callback, Pointer<Void> unlockCallbackFunc) async {
+    bool success = true;
+    try {
+      await callback();
+    } catch (error) {
+      success = false;
+      _realmLib.realm_register_user_code_callback_error(error.toPersistentHandle());
+    } finally {
+      _realmLib.realm_dart_invoke_unlock_callback(success, unlockCallbackFunc);
     }
   }
 
   static void _syncBeforeResetCallback(Object userdata, Pointer<shared_realm> realmHandle, Pointer<Void> unlockCallbackFunc) {
     final syncConfig = userdata as FlexibleSyncConfiguration;
-    final beforeResetCallback = syncConfig.clientResetHandler.onBeforeReset;
-    if (beforeResetCallback == null) {
-      _invokeNativeFunction(unlockCallbackFunc);
-      return;
-    }
+    final beforeResetCallback = syncConfig.clientResetHandler.onBeforeReset!;
+
     final realm = RealmInternal.getUnowned(syncConfig, RealmHandle._unowned(realmHandle));
-    _continueWhenComplete(() => beforeResetCallback(realm), unlockCallbackFunc);
+
+    _callResetCallback(() => beforeResetCallback(realm), unlockCallbackFunc);
   }
 
   static void _syncAfterResetCallback(Object userdata, Pointer<shared_realm> beforeHandle, Pointer<realm_thread_safe_reference> afterReference, bool didRecover,
       Pointer<Void> unlockCallbackFunc) {
     final syncConfig = userdata as FlexibleSyncConfiguration;
-    final afterResetCallback = didRecover ? syncConfig.clientResetHandler.onAfterRecovery : syncConfig.clientResetHandler.onAfterDiscard;
```

Comment:
Why is this check removed? You could still end up with a null callback if you do something like:


```dart
final handler = RecoveryOrDiscardHandler(onAfterRecovery: () {...});
```

If recovery fails, the `onAfterDiscard` needs to be invoked, but it'll be `null` here.

Reply:
which check was removed?

Reply:
The check for null

Reply:
the null check below 
```dart
if (afterResetCallback == null) {
```

was removed cause this funciton `_syncAfterResetCallback` now is only registered and invoked if either `onAfterRecovery` or `onAfterDiscard` is not null. 
hence the new implementation does a hard cast from null `onAfterRecovery!` and `onAfterDiscard!`

Is the question about these hard casts? Are they problematic?

Reply:
This check now is done here
https://github.com/realm/realm-dart/pull/1003/files#diff-cd0d6010ea23a618f899185aec525b9c186cc41c35679f48d21f2d295f5c8cfbR237

Reply:
As I said in my example - if `onRecoveryCallback` is not null, we're going to register: `config.clientResetHandler.onAfterRecovery != null || config.clientResetHandler.onAfterDiscard != null` returns `true`. But if `didRecover` is `false`, we'll try to execute `onAfterDiscard` which is `null`.

Reply:
handled